### ISSUE TITLE
Improve global container overview of the service

### ISSUE THREAD

**Issue by 2opremio** [weaveworks/scope#1032]
Here's a screenshot of the service with (only) three k8s worker nodes and not even a probe in the master node (because #1030 ):

<img width="1769" alt="screen shot 2016-02-25 at 12 42 18" src="https://cloud.githubusercontent.com/assets/2362916/13319964/4ae9c396-dbbd-11e5-903b-c4aed654df29.png">

As you can see, the edges are really tangled, plus the `uncontained` nodes don't tell us anything in the top-level view.

Report: [service_report.json.gz](https://github.com/weaveworks/scope/files/146491/service_report.json.gz)

We should try to improve the rendering (yes, the never-ending problem, I wish I could provide specific suggestions). 

I am not sure what we could do about the  `Uncontained` nodes. I think I'd rather see them expanded into processes nodes with names even if it's the container view. 

I think the container view makes the assumption that all your services/infrastructure is containerized, which may not be the case (like in kubernetes). 

For instance, in the case of k8s, all connections go through the kube-proxy. Seeing kube-proxy instead of `Uncontained` would be better (at least you see what's that `Unconnected` node)/ However, it's not good enough because  kube-proxy handles all the connections to services, which would result in a node like:

```
                  _______________________
container x ---> |                      | ---> container a
container y ---> |                      | ---> container b
container z ---> |      kube_proxy      | ---> container c
        ... ---> |                      | ---> ...
                  ----------------------
```

In which you have no idea where an origin connection is leading. We could make use of the kubernetes information (the service IPs and to what containers they map) to improve the view to something like this:

```
                        /---------> container a
conteainer x ---->service1 -------> container b
                        \---------> container c

                       /----------> container a
container y ---->service2 --------> container d
                       \----------> container e

```

(I didn't draw a single container a because it's complicated with ascii art). Or even:

```
_____________
             |---------> container a
container x  |----------> container b
             |---------> container c
-------------

_____________
             |---------> container a
container y  |----------> container d
             |---------> container e
-------------
```
(if we don't want to show services), which would be much more useful from the application's perspective.

But it would be good to be able to still see the infrastructural view (i.e. the kubelet_node) if wanted.

So to summarize I think we should:

* Improve rendering of crossed edges
* Show information about uncontained nodes at the top level
* Use semantic information to improve the view of the connections from and to the kube-proxy.

Thoughts?

**Comment by 2opremio** [comment]
The kube-proxy work is something we will need for flux too.

**Comment by 2opremio** [comment]
Related: https://github.com/weaveworks/scope/issues/606

**Comment by pidster** [comment]
Seems like we should treat the k8s processes as special system nodes, like we do with weave stuff.

**Comment by rade** [comment]
Related: #538

**Comment by 2opremio** [comment]
I've been doing some more research and it seems that the kube-proxy spaghetti ball will be no more after kubernetes 1.2, when kube-proxy will start using the iptables mode by default (see https://github.com/kubernetes/kubernetes/blob/master/docs/user-guide/services.md#virtual-ips-and-service-proxies )

My tests (in a 1-node cluster so far) show that the iptables mode plays well with conntrack. 

Kubernetes 1.2 will be released sometime in March (see https://github.com/kubernetes/kubernetes/wiki/Release-1.2). This makes eliding kube-proxy in userspace mode a much lower priority.

Our cluster service is running kubernetes 1.1.8, with kube-proxy running in userspace mode (resulting in the picture above). In order to elide kube-proxy accurately when running in userspace mode, we would need kube-proxy's port mapping information (i.e. the information printed at here https://github.com/kubernetes/kubernetes/blob/v1.1.1/pkg/proxy/userspace/proxysocket.go#L15 ) which is not publicly exposed. But, again, I am not sure this should be a priority anymore.

For reference here's my local cluster with kube-proxy running in userspace mode:

<img width="1766" alt="screen shot 2016-02-25 at 20 29 49" src="https://cloud.githubusercontent.com/assets/2362916/13333360/951e738a-dbfe-11e5-8ea2-c96fd4488e95.png">


and in iptables mode:

<img width="1768" alt="screen shot 2016-02-25 at 20 08 29" src="https://cloud.githubusercontent.com/assets/2362916/13332827/cab771a2-dbfb-11e5-8002-857af3e73b42.png">


I still need to make a test with a multinode cluster but I think Scope will work just fine with k8s>1.2

**Comment by 2opremio** [comment]
CC @monadic 

@pidster @rade  Thoughts? Should we bite the bullet and make Scope work with kube-proxy in userspace mode too?

**Comment by rade** [comment]
> Should we still bite the bullet and make Scope work with kube-proxy in userspace mode too?

As I mentioned on slack, we need figure out how to deal with proxies in general. There are plenty of other proxies around that people use, and some of them are in userspace by necessity, i.e. the kernel does not speak HTTP.

But there seems little point in investing in scoping userspace kube-proxy, if that is not what most people will be running soon.

**Comment by 2opremio** [comment]
> As I mentioned on slack, we need figure out how to deal with proxies in general. There are plenty of other proxies around that people use, and some of them are in userspace by necessity, i.e. the kernel does not speak HTTP.

I think that eliding users-pace proxies should be fairly easy once we have access to all the proxy's 4-tuples: `in.RemoteAddr(), in.LocalAddr(), out.LocalAddr(), out.RemoteAddr()`
 
The bigger problem will be getting that information.

**Comment by 2opremio** [comment]
Related #552

**Comment by 2opremio** [comment]
Related #361

**Comment by pidster** [comment]
+1, no point in investing in the user-space proxy.

**Comment by 2opremio** [comment]
> +1, no point in investing in the user-space proxy.

Uhm, maybe not specifically on userspace kube-proxy but this is a problem we are going to face in many other scenarios (flux, nginx, custom proxies ...).

With flux (and even kube-proxy), we can add introspection APIs which tell us to what backend(s) each proxy client specifically talking to. However that doesn't scale: we cannot modify every single userspace proxy solution out there (think nginx, custom user proxies we can't control etc ...) plus it goes against Weave's philosophy of _no instrumentation_.

I am thinking this could be a perfect opportunity to try to use the techniques in https://github.com/weaveworks/scope/tree/master/experimental/tracer driven by a specific usecase (trying to get an ingress->egress correlation in proxies without instrumentation).

Getting this to work will be hard (if not impossible) but, if we get it right, it would be a key differentiator with other monitoring solutions out there.

Thoughts? (CC @tomwilkie )

**Comment by tomwilkie** [comment]
Id say almost all decent proxies will use asynchronous io and make the
techniques in tracer useless.

If been keen to see if the layer 3 iptables based load balancing can be
intospected by scope. I can't see a way of doing the layer 7 work without
instrumentation. But in the whole one weave vein, this is definitely
something flux should do.
